pkg/services/statefetcher/statefetcher.go

```diff
@@ -0,0 +1,364 @@
+	}
+	containerMagic := containerObj.Attribute("Magic")
+	if containerMagic != strconv.Itoa(int(sfs.chain.GetConfig().Magic)) {
+		sfs.isShutdown.CompareAndSwap(false, true)
+		return nil, fmt.Errorf("container magic mismatch: expected %d, got %s", sfs.chain.GetConfig().Magic, containerMagic)
+	}
+	sfs.isShutdown.CompareAndSwap(true, false)
+	return &sfs, nil
+}
+
+// LastStateIndex returns the last state index from NeoFS.
+func (sfs *Service) LastStateIndex() (uint32, error) {
+	if sfs.lastStateIndex != 0 {
+		return sfs.lastStateIndex, nil
+	}
+	startIndex := uint32(1)
+	for {
+		prm := client.PrmObjectSearch{}
+		filters := object.NewSearchFilters()
+		filters.AddFilter(sfs.cfg.StateAttribute, fmt.Sprintf("%d", startIndex*sfs.stateSyncInterval), object.MatchStringEqual)
+		prm.SetFilters(filters)
+
+		ctx, cancel := context.WithTimeout(sfs.basic.Ctx, sfs.cfg.Timeout)
+		var (
+			blockOidsObject []oid.ID
+			err             error
+		)
+		err = sfs.basic.Retry(func() error {
+			blockOidsObject, err = neofs.ObjectSearch(ctx, sfs.basic.Pool, sfs.basic.Account.PrivateKey(), sfs.cfg.ContainerID, prm)
+			return err
+		})
+		cancel()
+		if err != nil {
+			sfs.isShutdown.CompareAndSwap(false, true)
+			return 0, fmt.Errorf("failed to find '%s' object with index %d: %w", sfs.cfg.StateAttribute, startIndex, err)
+		}
+		if len(blockOidsObject) == 0 {
+			sfs.lastStateIndex = (startIndex - 1) * sfs.stateSyncInterval
+			return sfs.lastStateIndex, nil
+		} else {
+			sfs.lastStateOid = blockOidsObject[0]
+		}
+		startIndex++
+	}
+}
+
+// Start begins state fetching.
+func (sfs *Service) Start() error {
+	if sfs.IsShutdown() {
+		return errors.New("service shut down")
+	}
+	if !sfs.isActive.CompareAndSwap(false, true) {
+		return nil
+	}
+	sfs.log.Info("state fetcher started")
+	sfs.wg.Add(1)
+	go sfs.run()
+	return nil
+}
+
+// Shutdown stops the service.
+func (sfs *Service) Shutdown() {
+	if !sfs.isActive.CompareAndSwap(true, false) {
+		return
+	}
+	sfs.basic.CtxCancel()
+	sfs.wg.Wait()
+
+	sfs.isShutdown.Store(true)
+	if sfs.shutdownCallback != nil {
+		sfs.shutdownCallback()
+	}
+	sfs.log.Info("state fetcher shutdown")
+}
+
+func (sfs *Service) run() {
+	var (
+		syncHeight   uint32
+		expectedRoot util.Uint256
+	)
+
+	defer func() {
+		sfs.isActive.Store(false)
+		sfs.isShutdown.Store(true)
+		sfs.basic.CtxCancel()
+		sfs.wg.Done()
+		if sfs.shutdownCallback != nil {
+			sfs.shutdownCallback()
+		}
+		sfs.log.Info("state fetcher stopped")
+	}()
+	reader, err := sfs.getStateObject()
+	if err != nil {
+		sfs.log.Error("failed to get state object", zap.Error(err))
+		return
+	}
+	defer func() {
+		if err = reader.Close(); err != nil {
+			sfs.log.Warn("failed to close reader", zap.Error(err))
+		}
+	}()
+
+	var (
+		lastKey = sfs.chain.GetLastStoredKey()
+		kvChan  = make(chan kvPair, 1)
+		batch   = make([]statesync.KeyValue, 0, sfs.cfg.KeyValueBatchSize)
+	)
+	go func() {
+		defer close(kvChan)
```

Comment:
This logic is too complicated and I don't quite understand the intention of making `kvChan` and allocating a separate routine. If you want to optimize the reading process then create a channel of KeyValue *batches* so that while single batch is being stored/persisted, the next batch is being read from NeoFS.